plugins/task-plugin/src/preview/previews-widget.ts

```diff
@@ -62,26 +66,31 @@ export class PreviewUrlsWidget {
                 </html>`;
     }
 
-    private render(): string {
+    private async render(): Promise<string> {
         if (this.options.tasks.length < 1) {
             return `<div class='previews-placeholder'>${PLACEHOLDER}</div>`;
         }
 
-        const previews = this.renderPreviews().join('');
+        const previews = (await this.renderPreviews()).join('');
         return `<div class='previews-container'>${previews}</div>`;
     }
 
-    private renderPreviews(): Array<string> {
-        return this.options.tasks.map(cheTask => {
-            const previewUrl = cheTask.definition.previewUrl;
-
-            const server = `<lable>${previewUrl}</label>`;
-            const taskLabel = `<lable>${cheTask.name}</label>`;
-            const previewButton = `<button class='button' type="button" onclick="preview('${INTERNALLY_CHOICE}', '${previewUrl}')">${PREVIEW_BUTTON_NAME}</button>`;
-            const goToButton = `<button class='button' type="button" onclick="preview('${EXTERNALLY_CHOICE}', '${previewUrl}')">${GO_TO_BUTTON_NAME}</button>`;
-
-            return this.renderTemplate(server, taskLabel, previewButton, goToButton);
-        });
+    private async renderPreviews(): Promise<Array<string>> {
+        return await Promise.all(
+            this.options.tasks.map(
+                async (cheTask: theia.Task) => {
+                    const previewUrl = cheTask.definition.previewUrl;
+                    const url = await this.previewUrlOpenService.resolve(previewUrl);
+
+                    const server = `<a class="task-link" href="${url}" target="_blank">${url}</a>`;
+                    const taskLabel = `<label>${cheTask.name}</label>`;
+                    const previewButton = `<button class='button' type="button" onclick="preview('${INTERNALLY_CHOICE}', '${previewUrl}')">${PREVIEW_BUTTON_NAME}</button>`;
+                    const goToButton = `<button class='button' type="button" onclick="preview('${EXTERNALLY_CHOICE}', '${previewUrl}')">${GO_TO_BUTTON_NAME}</button>`;
```

Comment:
If I understand  correctly, at this step you already have a resolved url, but unresolved url is passed to 'preview' function.
Do we have some reasons for it?
Maybe it makes sense to pass the resolved url, otherwise a url is resolved twice: [at rendering](https://github.com/eclipse/che-theia/pull/318/files#diff-76b3d37b7996f3b0708dbfc19bf23f41R83) and [when user click on 'Preview' or 'Go to' button](https://github.com/eclipse/che-theia/pull/318/files#diff-51367edb0f101b505665cc8cdd29af24R48) 
WDYT?

Reply:
Thank you @RomanNikitenko for pointing that out. That makes sense, I'll fix it.